### PR TITLE
Re-do developer tools service

### PR DESCRIPTION
**Pull request by bramkragten**
## Proposed change

image

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by balloob** [comment]
Just purely based on the screenshot: how do we deal with optional fields. Ie brightness doesn't have to be set but that's not something that we currently allow in our selector rendering.

**Comment by balloob** [comment]
Does it make sense to show the parameters table if users can't fill it in ? Maybe just a sentence "More parameters are available in YAML mode" if there are more fields available.

**Comment by frenck** [comment]
Yeah I agree with Paulus, the table doesn't make sense on this page and just showing it during YAML mode, would be sufficient.

As for optional fields, I think I'll drop my checkbox suggestion again 😁

**Comment by bramkragten** [comment]
I changed the table yesterday to only show the advanced parameters in UI mode, and all + target is YAML mode.

For optional/required, I wanted to show required with the usual star or something and check and notify the user when he presses call service, but yeah Frenck and I had some discussions about it already.

One option would be to add a checkbox before every optional parameter, that would also fix the issue of clearing an optional `select` for instance.

**Comment by balloob** [comment]
Let's stick with the checkbox. Ideally we would disable the selector when it's disabled so it's clear the value won't be added.

File: src/components/ha-service-control.ts [hunk 1]
```diff
@@ -48,28 +50,60 @@ export class HaServiceControl extends LitElement {
...
+        target,
+        data: { ...this.value.data },
       };
+
       delete this.value.data!.entity_id;
+      delete this.value.data!.device_id;
+      delete this.value.data!.area_id;
     }
 
     if (this.value?.data) {
```
Comment (ludeeus) [inline src/components/ha-service-control.ts:109]: At this point we know `this.value` has content, the `?` is not needed

File: src/components/ha-yaml-editor.ts [hunk 2]
```diff
@@ -44,14 +44,14 @@ export class HaYamlEditor extends LitElement {
...
+  @query("ha-code-editor") private _editor?: HaCodeEditor;
 
   public setValue(value): void {
     try {
       this._yaml = value && !isEmpty(value) ? safeDump(value) : "";
     } catch (err) {
       // eslint-disable-next-line no-console
-      console.error(err);
+      console.error(err, value);
       alert(`There was an error converting to YAML: ${err}`);
```
Comment (ludeeus) [inline src/components/ha-yaml-editor.ts:55]: Should be converted to `showAlertDialog` with localize and include the value in it, and remove the console statement

File: src/panels/developer-tools/service/developer-tools-service.ts [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+import "../../../components/ha-yaml-editor";
+import "../../../components/ha-service-picker";
+import "../../../styles/polymer-ha-style";
+import { HomeAssistant } from "../../../types";
+import "../../../util/app-localstorage-document";
+import "../../../components/ha-service-control";
+import { computeDomain } from "../../../common/entity/compute_domain";
+import { computeObjectId } from "../../../common/entity/compute_object_id";
+import { haStyle } from "../../../resources/styles";
+import memoizeOne from "memoize-one";
```
Comment (ludeeus) [inline src/panels/developer-tools/service/developer-tools-service.ts:22]: sort imports?

File: src/panels/developer-tools/service/developer-tools-service.ts [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+              @value-changed=${this._yamlChanged}
+            ></ha-yaml-editor>`
+          : html`<ha-service-control
+              .hass=${this.hass}
+              .value=${this._serviceData}
+              showAdvanced
+              @value-changed=${this._serviceChanged}
+            ></ha-service-control>`}
+
+        <div class="button-row">
```
Comment (ludeeus) [inline src/panels/developer-tools/service/developer-tools-service.ts:65]: Should these be visible if no service is selected?

Can we use `display: flex;justify-content: space-between;` on the button-row class to move YAML to the right and out of the main focus?

**Review by ludeeus — APPROVED** [review]

File: src/components/ha-service-control.ts (hunk 1 above)
Comment (bramkragten) [reply src/components/ha-service-control.ts:109]: No, `value` could be undefined here? All the checks above are in a different if statement.

File: src/components/ha-yaml-editor.ts (hunk 2 above)
Comment (bramkragten) [reply src/components/ha-yaml-editor.ts:55]: I'll do a follow up PR.

File: src/components/ha-service-control.ts (hunk 1 above)
Comment (ludeeus) [reply src/components/ha-service-control.ts:109]: You set it with  `data`on line 98

**Comment by balloob** [comment]
From the screenshot: feels weird that the dropdown stretches the full width of the page.

**Comment by bramkragten** [comment]
> From the screenshot: feels weird that the dropdown stretches the full width of the page. 

Why?

**Comment by balloob** [comment]
Ah didn't realize we showed that. Even then I much prefer if that wasn't that long. Difficult to read long sentences.

**Comment by frenck** [comment]
I think we should swap them maybe. Show the service name mainly and the description underneath. We are now using a description as a title/name.

**Comment by bramkragten** [comment]
Then we should add a name/title, i think the service name is too technical...

**Comment by frenck** [comment]
yeah a name, just like the fields would make sense

File: src/components/ha-service-control.ts (hunk 1 above)
Comment (bramkragten) [reply src/components/ha-service-control.ts:109]: But that is in the if block from line 77?

Comment (ludeeus) [reply src/components/ha-service-control.ts:109]: Ah, right!

File: src/panels/developer-tools/service/developer-tools-service.ts [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+                : this.hass.localize(
+                    "ui.panel.developer-tools.tabs.services.yaml_parameters"
+                  )}
+              outlined
+              .expanded=${this._yamlMode}
+            >
+              ${this._yamlMode && target
+                ? html`<h3>
+                    ${this.hass.localize(
+                      "ui.panel.developer-tools.tabs.services.accepts_target"
```
Comment (ludeeus) [inline src/panels/developer-tools/service/developer-tools-service.ts:108]: Should we use a markdown element to render this?

> This service accepts a target, for example: \`entity_id: light.bed_light\`

kinda looks broken

**Comment by balloob** [comment]
When I open dev tools for the first time and no service is selected, and I click on YAML mode, I get this:

image

That might be confusing. Could we fill in "basic" fields like `service` and `data` ?

**Comment by balloob** [comment]
The checkboxes being above the field on mobile takes up a lot of space. Maybe on mobile we can hide the description behind a help icon that pops it up and put the checkbox before the title.

image

**Comment by balloob** [comment]
The description maybe should always be behind a help button, as it can get really cramped.

image

**Comment by balloob** [comment]
We should come up with style guidelines for services.yaml. For example, writing down the possible values should no longer be part of the description.

image

**Comment by balloob** [comment]
Oh I just got an idea, can we hide the controls when an option is disabled? That way it won't take up a lot of space.

**Comment by balloob** [comment]
And if it's disabled, we can show the description instead of the controls. And then when control is visible, the description is behind a help icon.

**Comment by balloob** [comment]
The problem with rendering titles like this while sorting by the service name, is that the list ordering makes no sense. There are a lot of services so this will be annoying to browse. Maybe we can prefix each service name with the domain and then sort on the service title.

 - Light: Turn on a light
 - Light: Turn off a light
 - Light: Toggle a light
 - Switch: Turn on a switch
 - etc

image

**Pull request merged**